components/script/dom/idbfactory.rs

```diff
@@ -0,0 +1,227 @@
+                }
+            },
+        };
+
+        if new_version == db_version {
+            upgrade_notificator.send(Ok(())).unwrap();
+            result_channel.send(Ok(Trusted::new(&result))).unwrap();
+        } else if new_version > db_version {
+            // Step 5.10.4
+            // FIXME:(rasviitanen) Do this for all open handles
+            result.dispatch_versionchange(db_version, version);
+
+            // Step 5.10.6
+            // Dispatch an event to trigger version change
+            let transaction = IDBTransaction::new(
+                &self.global().as_window(),
+                &result,
+                IDBTransactionMode::Versionchange,
+                result.object_stores(),
+            );
+
+            transaction.set_active_flag(false);
+            transaction.start_and_upgrade_version(new_version); // Starts transaction
+
+            result.set_transaction(&transaction);
+
+            request.dispatch_upgrade_transaction(
+                &result,
+                &transaction,
+                db_version,
+                Some(new_version),
+                upgrade_notificator,
+            );
+
+            result_channel.send(Ok(Trusted::new(&result))).unwrap();
+        } else {
+            upgrade_notificator.send(Ok(())).unwrap();
+            result_channel.send(Err(Error::Version)).unwrap();
+        }
+    }
+}
+
+impl IDBFactoryMethods for IDBFactory {
+    // https://www.w3.org/TR/IndexedDB-2/#dom-idbfactory-open
+    #[allow(unsafe_code)]
+    fn Open(&self, name: DOMString, version: Option<u64>) -> Fallible<DomRoot<IDBOpenDBRequest>> {
+        // Step 1: If version is 0 (zero), throw a TypeError.
+        if version == Some(0) {
+            return Err(Error::Type(
+                "The version must be an integer >= 1".to_owned(),
+            ));
+        };
+
+        // Step 2: Let origin be the origin of the global scope used to
+        // access this IDBFactory.
+        let global = self.global();
+        let origin = global.origin();
+        let window = global.as_window();
+
+        // Step 3: if origin is an opaque origin,
+        // throw a "SecurityError" DOMException and abort these steps.
+        if let ImmutableOrigin::Opaque(_) = origin.immutable() {
+            return Err(Error::Security);
+        }
+
+        // Step 4: Let request be a new open request.
+        let request = IDBOpenDBRequest::new(&self.global().as_window());
+
+        // Step 5: Runs in parallel
+        let (task_source, canceller) = window
+            .task_manager()
+            .dom_manipulation_task_source_with_canceller();
+
+        let this = Trusted::new(self);
+        let trusted_request = Trusted::new(&*request);
+        let name_as_string = name.to_string(); // DOMString is not Send
+        Builder::new()
+            .name("IndexedDBOpenDatabase".to_owned())
+            .spawn(move || {
+                // FIXME:(rasviitanen) This should not be queued as a task,
```

Comment:
Ah ok so I understand a bit better why you use a thread here, and I think you want to restructure this a bit. 

In general when the spec says to do something "in parallel", in this context it probably means doing it somewhere inside `net`. Then you want to setup a route using the `ROUTER` to get the result from that parallel operation over IPC, and then queue a task from the router thread to the event-loop to have that result affect the world of javascript. 

So what you're doing here is running `open_internal` on the event-loop, effectively blocking it while waiting for the IPC reply from `net`, and then you message back to this thread. This thread is indeed running "in-parallel" to the event-loop, however the blocking `create_database` task is then run on the event-loop which defeats the purpose of "in-parallel" steps. 

So everytime you do a `recv()` on an ipc channel here in `script`, I think you actually want to use the `ROUTER`, do the "in-parallel" stuff back in `net`, and then reply back to the route, and finally queue a task with the result(Unless the spec really means to make an operation blocking)